common/transport/amqp/src/receiver_link.ts

```diff
@@ -21,57 +28,119 @@ const debug = dbg('azure-iot-amqp-base:ReceiverLink');
 
-  constructor(linkAddress: string, linkOptions: any, amqp10Client: any) {
+  constructor(linkAddress: string, linkOptions: ReceiverOptions, session: Session) {
     super();
     this._linkAddress = linkAddress;
     this._linkOptions = linkOptions;
-    this._amqp10Client = amqp10Client;
+    this._rheaSession = session;
+    this._unDisposedDeliveries = [];
+    this._combinedOptions = {
+      source: linkAddress
+    };
 
-    this._detachHandler = (detachEvent: any): void => {
-      debug('handling detach event: ' + JSON.stringify(detachEvent));
-      this._fsm.handle('forceDetach', detachEvent.error);
+    if (linkOptions) {
+      for (let k in linkOptions) {
+        this._combinedOptions[k] = linkOptions[k];
+      }
+    }
+
+    const receiverOpenHandler = (context: EventContext): void => {
+      this._fsm.handle('receiverOpenEvent', context);
+    };
+    const receiverCloseHandler = (context: EventContext): void => {
+      this._fsm.handle('receiverCloseEvent', context);
+    };
+    const receiverErrorHandler  = (context: EventContext): void => {
+        this._fsm.handle('receiverErrorEvent', context);
     };
 
-    this._errorHandler = (err: Error): void => {
-      debug('handling error event: ' + err.toString());
-      this._fsm.handle('forceDetach', err);
+    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_012: [If a `message` event is emitted by the `rhea` link object, the `ReceiverLink` object shall emit a `message` event with the same content.]*/
+    const receiverMessageHandler = (context: EventContext): void => {
+      this._unDisposedDeliveries.push({msg: context.message as any, delivery: context.delivery});
+      this.emit('message', context.message);
     };
 
-    /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_012: [If a `message` event is emitted by the `amqp10` link object, the `ReceiverLink` object shall emit a `message` event with the same content.]*/
-    this._messageHandler = (message: AmqpMessage): void => {
-      this.emit('message', message);
+    const manageReceiverHandlers = (operation: string) => {
+      this._rheaReceiver[operation]('receiver_error', receiverErrorHandler);
+      this._rheaReceiver[operation]('receiver_close', receiverCloseHandler);
+      this._rheaReceiver[operation]('receiver_open', receiverOpenHandler);
+      this._rheaReceiver[operation]('message', receiverMessageHandler);
     };
 
     this._fsm = new machina.Fsm({
+      /*Codes_SRS_NODE_AMQP_RECEIVER_LINK_16_001: [The `ReceiverLink` internal state machine shall be initialized in the `detached` state.]*/
       initialState: 'detached',
-      namespace: 'receiverlink',
+      namespace: 'receiverLink',
       states: {
         detached: {
           _onEnter: (callback, err) => {
-            this._linkObject = null;
+            this._rheaReceiver = null;
+            this._rheaReceiverName = null;
 
             if (callback) {
-              callback(err);
+              this._safeCallback(callback,err);
```

Comment:
>,err [](start = 41, length = 4)

nit: spacing. also: what's the use of _safeCallback here?

Reply:
it seems you're using safeCallback almost everywhere. it blows up the stack and ruins the debugging experience, also makes writing unit tests much harder. why is it necessary to nextTick all that stuff? or to test if a callback is there or not, considering it should always be there? in this specific case, you're even going to test if the callback is not undefined twice before calling it


---
In reply to: [209352713](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209352713) [](ancestors = 209352713)